### PR TITLE
Replace chalk with nanocolors

### PR DESCRIPTION
**Pull request by ai**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | 👍
| License                  | MIT

I color output library from `chalk` to [Nano Colors](https://github.com/ai/nanocolors). Reasons:

- Nano Colors has no dependencies and use only **[16 kB](https://packagephobia.com/result?p=nanocolors)** in `node_modules`. `chalk` has [5 dependencies](https://npm.anvaka.com/#/view/2d/chalk) and takes **[101 kB](https://packagephobia.com/result?p=chalk)**.
- Nano Colors is [faster](https://github.com/ai/nanocolors#benchmarks). It **7 times** faster for loading and **5 times** faster for method calls.
- Nano Colors is already used in **Autoprefixer, Browserslist, PostCSS**, and Size Limit. [>800 000 downloads](https://npm-stat.com/charts.html?package=nanocolors) per day.
- Nano Colors supports **Node.js ≥ 6** ([CI](https://github.com/ai/nanocolors/blob/main/.github/workflows/test.yml#L41-L58)) and both ESM and CJS imports (even universal browsers/Node.js projects).
- Nano Colors has nice tree-shakable API: `import { red } from 'nanocolors'`.
- Color support detection uses Node.js `tty`, `TERM`, `NO_COLOR`, `FORCE_COLOR` env variables. It is ready for Windows and CI.
- It will have a good support. My other projects from nano-series like very popular [Nano ID](https://github.com/ai/nanoid/) shows a minimal number of bugs and excellent response time.

### CODE REVIEW

**Comment by ai** [comment]
Seems like I need help with E2E tests, since they are out of this repo:

1. `E2E (jest)` is falling of a little different color code points order. They have no visual changes, just a different way to encode the same output (for performance reasons).
2. `E2E (babel-old-version)` is falling because of some Yarn dependency issue. Have no idea how to fix it because I am not fully understand `package.json` tricks in `scripts/integration-tests/e2e-babel-old-version.sh`.

**Comment by nicolo-ribaudo** [comment]
2. has just been fixed on the `main` branch

**Comment by ai** [comment]
@nicolo-ribaudo thanks for the advice! I fixed both issues.

Just have an issue between Jest and dual CJS/ESM packages like `nanocolors` (it takes ESM file, but we have CJS file mentioned in `package.main`). I will think today, but maybe you saw it before?

**Comment by nicolo-ribaudo** [comment]
I can take a look at it; we have a custom Jest resolver that might cause problems.

**Comment by nicolo-ribaudo** [comment]
Ugh, I cannot reproduce the failure locally :confused:

**Comment by ai** [comment]
> Ugh, I cannot reproduce the failure locally confused

Same for me :-/

I am planning to check Node.js version or randomly change resolver (like adding `.cjs` extension).

**Comment by nicolo-ribaudo** [comment]
Ok I can reproduce it running `./scripts/integration-tests/publish-local.sh` and `./scripts/integration-tests/e2e-babel-old-version.sh`
:warning: If you run these command, they will `yarn global add verdaccio-memory@~10.0.0` and if you kill the scripts they'll change your `git config user.name` and `git config user.email`.

**Comment by nicolo-ribaudo** [comment]
@SimenB I don't know if it's expected, but this is another case where the `conditions` passed to the resolver are undefined.

I added
```js
  if (request.includes("nano")) console.log("RESOLVE", request, options);
```
to https://github.com/babel/babel/blob/6818b22dc0b1defcf39f6b876070d9ec5f8c4e46/test/jestExportsMapResolver.cjs#L28, and it logs:
```
RESOLVE nanocolors {
  basedir: '/home/nicolo/Documenti/dev/babel/babel-e2e-old/node_modules/jest-message-util/node_modules/@babel/highlight/lib',
  browser: undefined,
  conditions: undefined,
  defaultResolver: [Function: defaultResolver],
  extensions: [ '.js', '.jsx', '.ts', '.tsx', '.json', '.node' ],
  moduleDirectory: [ 'node_modules' ],
  paths: undefined,
  rootDir: '/home/nicolo/Documenti/dev/babel/babel-e2e-old'
}
```

I think this happens when loading the `@babel/highlight` used as a dependency of Jest, not when loading the `@babel/highlight` from our monorepo in the tests. Our E2E test fails because we publish the new `@babel/highlight` version which depends on `nanocolors`, and Jest depends on it.

I can easily work around the problem, but I'd like to know if it's expected.

---

@ai To make the tests pass, you can replace this line:
```diff
- const resolver = getResolver(options.conditions || ["default"]);
+ const resolver = getResolver(options.conditions || ["require", "default"]);
```
https://github.com/babel/babel/blob/6818b22dc0b1defcf39f6b876070d9ec5f8c4e46/test/jestExportsMapResolver.cjs#L28

**Review by nicolo-ribaudo — APPROVED** [review]
I usually dislike adding a dependency on a very young package, but given that it's maintained by @ai (who also maintains browserslist) the size difference ([20.4 kB](https://bundlephobia.com/package/chalk@4.1.2) vs [1.6 kB](https://bundlephobia.com/package/nanocolors@0.2.3)) and the perf improvement are worth it. 

Thank you @ai!

**Comment by jorgebucaran** [comment]
Dear Babel maintainers, thank you for all your work. @nicolo-ribaudo 

I strongly advise against migrating to nanocolors. Let me explain.

@ai, who is well-known in the JavaScript community for Autoprefixer, PostCSS, etc., shamelessly copied one of my most downloaded node packages ([Colorette](https://github.com/jorgebucaran/colorette), >20M/week) and rebranded it as nanocolors. 

A substantial amount of work and hours have gone into Colorette over the years, we've fixed numerous bugs and found creative ways to optimize performance, decrease package size, and more. nanocolors blatantly plagiarizes all this work. It's unethical and unprofessional.

Seeing him leverage his notability and following to promote and increase adoption of nanocolors, which he just released a few days ago, is unethical and disgraceful. As an open-source maintainer myself I feel discouraged. 

This is not in the true spirit of open source.

**Review by jorgebucaran — CHANGES_REQUESTED** [review]
Dear Babel maintainers, thank you for all your work. @nicolo-ribaudo 

I strongly advise against migrating to nanocolors. Let me explain.

@ai, who is well-known in the JavaScript community for Autoprefixer, PostCSS, etc., shamelessly copied one of my most downloaded node packages ([Colorette](https://github.com/jorgebucaran/colorette), >20M/week) and rebranded it as nanocolors. 

A substantial amount of work and hours have gone into Colorette over the years, we've fixed numerous bugs and found creative ways to optimize performance, decrease package size, and more. nanocolors blatantly plagiarizes all this work. It's unethical and unprofessional.

Seeing him leverage his notability and following to promote and increase the adoption of nanocolors (https://github.com/eslint/eslint/pull/15102), which he just released a few days ago, is unethical and disgraceful. As an open-source maintainer myself I feel profundly discouraged. 

This is not in the true spirit of open source.

**Comment by SimenB** [comment]
@nicolo-ribaudo could you add a `new Error().stack` so we can see where the call comes from?

**Comment by ai** [comment]
@jorgebucaran sorry for PostCSS/Browserslist migration from `colorette`. I understand your feeling (but it was an only move in case of `colorette` 2.x API changes and following API breaking change in patch release).

If you want to send another PR and replace `chalk` to `colorette`, I am OK with it. Babel’s users will be happy in both scenarios, since all `colorette` and `nanocolors` are both much smaller than `chalk`.

> rebranded it as nanocolors

It is not true. Nano Colors was influenced by `colorette` API and internal design and I mentioned it in top section of Nano Colors docs. However, it is not rebranding:

1. I mix ideas not only from `colorette` but also from [`kleur`](https://github.com/lukeed/kleur), another great color formatting library.
2. `colorette` and `nanocolors` has very different results in [benchmark](https://github.com/ai/nanocolors#benchmarks) showing difference in internal implementation.
3. If you compare [`kleur`](https://github.com/lukeed/kleur/blob/master/colors.mjs), [`nanocolors`](https://github.com/ai/nanocolors/blob/main/index.js), and [`colorette`](https://github.com/jorgebucaran/colorette/blob/main/index.js) you will see many similarities (because it is open source, and they share ideas between each other), but also many differences:
  * In color detection. Nano Colors uses `process.argv` to be compatible with Chalk.
  * In color’s functions. Nano Colors do not generate RegExp on every call for performance reasons. It also doesn’t have `init`/`raw` separation.

> This is not in the true spirit of open source.

Forks and idea sharing are also an important part of open-source if they are coming with respect to authors of origin ideas and mentioning them.

Before `colorette 2.x` API changes, I promote your project to other projects because I tried to find `chalk` alternative. I sent you PRs. I used your library in all my projects for a year.

We all stand on the shoulders of giants. We both used Chalk color detection algorithm as reference. We both using other source code to find the best optimizations.

If you do not like that your tools is also used as an inspiration (with mentioning you), feel free to replace this PR with PR of replacing `chalk` to `colorette`.

**Comment by mindplay-dk** [comment]
@ai any particular reason you "soft forked" and rebranded the original package? The differences seem ​minor and probably could have been a pull request?

I know you credited the original author, but now you're out here replacing his package with your rebranded package, making his work all but invisible.

I know this is open source, and maybe that's just the name of the game, but it doesn't seem very respectful and it could definitely look like you're trying to steal his fire.

If the package had been unmaintained, or your PR had been rejected, whatever - but you could have tried contributing before appropriating, could you not?

That would seem more in the spirit of collaboration and open source.

**Comment by ai** [comment]
> If the package had been unmaintained, or your PR had been rejected, whatever - but you could have tried contributing before appropriating, could you not? 

I did contribute to `colorette`. For I year I used it in my projects and promote it in Twitter.

The fork was created because @jorgebucaran changed `colorette` 2.x API (default exports stopped to use color support auto-detection, which made it not very usable in most popular use cases). Users (from webpack, for instance) ask him to revert changes bug, he told that this is how he see a project.

What I was able to do in that case? I forked a project and moved my projects to the fork.

When @jorgebucaran found that he lost `colorette` main users, he reverted 2.x API changes, but did it in patch release. Changes API in patch release was lost my trust in him as a maintainer.

**Comment by ai** [comment]
If you want to replace chalk to project without drama, I can recommend [`kleur`](https://github.com/lukeed/kleur). It has the same performance with Nano Colors, the same ESM-first design in `kleur/colors` and also well maintained.

**Comment by arestov** [comment]
@jorgebucaran can you please suggest what @ai could fix (in addition to everything mentioned above https://github.com/babel/babel/pull/13783#issuecomment-927131266 and compatibility with chalk), so nanocolors will become from "shamelessly copied from colorette" to fair competitor fork?

**Comment by jorgebucaran** [comment]
I came here to call you out on what I consider unethical behavior. That's it. That actually led to several good improvements, like adding Colorette's git history to nanocolors. I thank you for that. Had I not confronted you or spoken out, I doubt any of this would've materialized. 💯 

Now, I believe we could've handled this much better. You too could've asked me before deciding to roll your own Colorette and migrate all your projects without a warning, aggressively promoting it, and sending PRs here and there. I would've been delighted to discuss anything. 

I don't like the skewed version of the story you are painting. I don't like that I'm hijacking this thread either. 

Thank you all for entertaining my views on this issue.

**Comment by jorgebucaran** [comment]
One more thing. I suggest running your benchmarks in a workflow. It's more reliable and anyone can easily verify it.

https://github.com/jorgebucaran/color-bench-test/runs/3710045880?check_suite_focus=true

Screen Shot 2021-09-26 at 7 27 24

**Comment by ai** [comment]
> I suggest running your benchmarks in a workflow

@jorgebucaran glad that you backported Nano Colors optimization. I hope you will mention inspiration from Nano Colors in your docs too.

**Comment by arestov** [comment]
1. https://github.com/lukeed/kleur/commit/41e46965738fca5b20e8b7e2e389e812b3200407 just 3 days later https://github.com/jorgebucaran/colorette/commit/0226f478e6ba8ac3060187fdf29383314900f072 https://github.com/jorgebucaran/colorette/pull/16
2. https://github.com/ai/nanocolors/commit/f421c2cc37504b51afd310ef297b52660327da3f  5 hours later https://github.com/jorgebucaran/colorette/commit/b1d82b9a9204bdac19b6306c2426a179f4cd6c6b 

@jorgebucaran is there anything that can be done, so you would stop calling that nanocolors was unethically, "shamelessly copied", "blatantly plagiarized" from colorette?

**Comment by jorgebucaran** [comment]
Sorry to hijack this thread again, but I've been invoked. 😞  

I didn't backport anything from nanocolors, on the contrary, I [showed you how you had broken color functions](https://github.com/jorgebucaran/color-bench-test/pull/1#issuecomment-927139787), and [you even thanked me](https://github.com/jorgebucaran/color-bench-test/pull/1#issuecomment-927141937). Obviously, `!s` avoids a runtime error. 



I went further and pointed out other errors to you too. If I had not done that, you'd have kept nanocolors broken for valid non-string types like numbers and dates. It's puzzling that you had to be convinced to fix it in the first place. 🤦‍♂️ 



I even opened an issue in nanocolors for a bug I once had too (https://github.com/ai/nanocolors/issues/17), which you introduced by fiddling with color initialization, and explained how it could be fixed. I did that in good faith. 



I thought this situation was settled when you included my copyright notice in your repository, updated your README, and retroactively modified the `git` history to include Colorette's, for which [I thanked you above](https://github.com/babel/babel/pull/13783#issuecomment-927173279). But the fact is, I shouldn't have had to call you out to achieve that, you should've come up with that yourself. 

I noticed that [avoiding an explicit `s === undefined`](https://github.com/jorgebucaran/colorette/commit/b1d82b9a9204bdac19b6306c2426a179f4cd6c6b#) boosts perf for most cases. That's the point of `s === ""`, and `!s` before that. I didn't borrow it from you. You are soft-bullying me into saying you came up with _a_ `!s` and that Colorette was somehow inspired by nanocolors too (https://github.com/jorgebucaran/colorette/issues/78)? Sounds like something pulled out of [@horse_js](https://twitter.com/horse_js). 



I'm willing to put our differences behind us. The community doesn't really need two practically identical packages that accomplish the same thing. I'd be happy to add you as a Colorette maintainer and shake hands on this. How about we work together instead? 

---

@arestov [Here](https://github.com/babel/babel/pull/13783#issuecomment-927173279). Calling him out is what led to all the nice improvements, like adding Colorette's git history to nanocolors and the COPYRIGHT file in the first place.

**Comment by mindplay-dk** [comment]
This whole "who copied who" debacle aside, if I were to review the PR itself, I guess I would say, with my conservative mindset:

Are there any important reasons to switch from one package to another? They are identical on the surface.

Would the end user notice anything different? Reasonably, no.

Would the maintainers notice any difference? No.

The current package works, is maintained and supported, is used by far more packages, and is far more popular.

This new package is brand new, doesn't even have a 1.x tag yet, and the author of the original package actually pointed out several bugs (that were fixed) just since this PR was opened.

There's no pressing or compelling reason to switch, as far as I can tell.

My conservative advice would be, since this doesn't solve any problem, I would stick with what already works.

**Comment by nicolo-ribaudo** [comment]
Hi everyone,
since this ethics discussion is about something that is happening right now, we don't feel comfortable with making a decision about this PR immediately. **We will wait two weeks before possibly merging it**, unless @ai and @jorgebucaran agree on what to do.

I personally don't have an opinion (yet?) on what's better for Babel between nanocolor, colorette and kleur. However, using _any_ of them looks like an improvement over `chalk`, which is what Babel is currently using.

There will probably still be some activity in this PR to discuss about https://github.com/babel/babel/pull/13783#issuecomment-927097401 (@SimenB I promise your stakc trace request wasn't overlooked between all the other comments, I'll post it later!), but please keep the discussion about the "nanocolors vs colorette" out of this repository, unless it's for technical stuff that would affect how they are used by Babel.

**Comment by arestov** [comment]
> Calling him out is what led to all the nice improvements

@jorgebucaran does it mean that the ethical part was solved? and from the ethical part (but not technical, maintenance, etc.) you are ok to accept merge?

**Comment by jorgebucaran** [comment]
I will refrain from further comments as requested by Nicolò. Feel free to reach out to me elsewhere.

**Comment by sindresorhus** [comment]
I'm one of the maintainers of Chalk. I don't really care what Babel uses, but I just thought I would point of some things:

---

> Nano Colors has no dependencies and use only 16 kB in node_modules. chalk has 5 dependencies and takes 101 kB.

The next version of Chalk will have no dependencies too and will be smaller (although not as small as 16 kB).

Also, Chalk is such a common dependency that it's usually already somewhere in the dependency tree on most projects anyway. So Chalk is often 0 kB.

And it's a bit unfair to compare sizes like this without any context as Chalk does a lot more (chainable API, 16 million color support, handles many edge-cases, etc) and also has TypeScript types.

> Nano Colors is faster. It 7 times faster for loading and 5 times faster for method calls.

These are meaningless numbers. Both Nano Colors and Chalk are more than fast enough. This is just micro-benchmarking (which does not reflect real-world usage) for the sake of marketing. Of course Nano Colors will be a little bit faster in a micro-benchmark since it has less code.

> Nano Colors is already used in SVGO, webpack-dev-server, Autoprefixer, Browserslist, PostCSS, and Size Limit. >800 000 downloads per day.

4 of these are your own projects. Chalk has 15 million downloads per day.

> Nano Colors has nice tree-shakable API: import { red } from 'nanocolors'.

This is not very relevant for a Node.js-based command-line coloring package.

I would also argue that's a bad use of named imports as it's a chore to import each color manually, you litter the scope with generic names, and it makes it less clear at the call site what, for example, `dim()` actually does (without having to scroll up and look at the import).

> Color support detection uses Node.js tty, TERM, NO_COLOR, FORCE_COLOR env variables. It is ready for Windows and CI.

Chalk suoports this too. Actually, most of these "tiny" color packages have	borrowed the color detection logic from Chalk.

> It will have a good support. My other projects from nano-series like very popular Nano ID shows a minimal number of bugs and excellent response time.

Chalk has had good maintenance and support for 8 years.

---

In short, I'm not arguing for or against merging this. Do what's right for the Babel project, but do it for the right reasons.

Chalk is mature, feature rich, and more than fast enough.

**Comment by Qix-** [comment]
Another Chalk maintainer here. I agree with everything @sindresorhus says. If this is genuinely better for the Babel project, then great. But the listed reasons are a bit disingenuous indeed, especially given the amount of work that's gone into Chalk over the years.

At least from my point of view, it's been kind of an unspoken nicety not to push one's own new packages to new software for altruistic reasons. It's something we've seen quite a lot over the years and it rarely improves the project in a meaningful way. I don't believe this case to be any different, personally.

If there is something that is negatively impacting Babel stemming from its use of Chalk, please by all means let us know - as Sindre pointed out, we've been active maintainers for almost a decade at this point. We're always open to ideas for improvement.

**Comment by pyros2097** [comment]
These things happen a lot in open source where someone forks/copies most of another project with slight changes and that becomes famous see [WebReflection on lit-html](https://gist.github.com/WebReflection/ab43649d9e4a53ac900b5924c77a310e). But I would prefer to have chalk in here anyways since the others are not solving an immediate problem as of such.

**Comment by 19h** [comment]
If a new dependency reduces the overall amount of dependencies while introducing non-trivial performance improvements then by all means it should be seriously considered independent of how old and maintained the replaced dependency is.

Babel is used by hundreds of thousands of developers, if not millions, and there should always be competition to improve performance -- a single second of improvement translates into millions of seconds of total wasted time.

**Comment by ai** [comment]
> The next version of Chalk will have no dependencies too and will be smaller (although not as small as 16 kB).

Wow. The `chalk` performance problems is the main reason why I moved from chalk` a year ago (to colorette 😅).

Any ideas of npm package size and performance of chalk 5?

You can copy this [performance](https://github.com/ai/nanocolors/commit/8edabdac6b29c9e4fbf2ad860c409ab5775f8280) [optimization](https://github.com/ai/nanocolors/commit/f207243be57af96767c9f32815f31fd9f390c667) ideas.

I would like for `chalk` to be competitive with new color libraries in terms of performance. I know that you think that it is just a number, but they are really matters on high scale. Performance parity will make it unnecessary for me to support another library.

**Comment by sindresorhus** [comment]
> Babel is used by hundreds of thousands of developers, if not millions, and there should always be competition to improve performance -- a single second of improvement translates into millions of seconds of total wasted time.

Babel is having a lot of competition where it matters (SWC, etc); parse and compile performance. Saving a few nanoseconds (pun intended) when printing the terminal output is not saving anyone any time.

**Comment by 19h** [comment]
@sindresorhus you‘re absolutely right. Still if this same conversation would be had about any other dependency we’d get nowhere. This conversation most definitely sparked optimizations in both colorette and chalk — why did this not happen before? Competition is absolutely important.

**Comment by sindresorhus** [comment]
> The chalk performance problems is the main reason why I moved from chalk` a year ago (to colorette 😅).

No one has complained to us about Chalk's performance since we fixed it in [Chalk v3](https://github.com/chalk/chalk/releases/tag/v3.0.0) (2 years ago), but we are always open to working on improving performance.

> Any ideas of npm package size and performance of chalk 5?

I estimate around 20 kB package size and 40 kB install size (32 kB if you exclude the TypeScript types). However, this measurement is flawed too. We could just move all the readme contents to a docs folder (which is not inclduded in the package) to save some kilobytes.

Performance like now or better if we manage to optimize for micro-benchmarks.

> I would like for chalk to be competitive with new color libraries in terms of performance. I know that you think that it is just a number, but they are really matters on high scale.

I'm curious what kind of high scales you're using Nano Colors on where it matters whether it runs 11K or 47K operations a second? And that's just the simple micro-benchmark (which again does not reflect real world usage). In your [complex benchmark](https://github.com/ai/nanocolors#benchmarks), Chalk is much closer.

**Comment by ai** [comment]
> I'm curious what kind of high scales you're using Nano Colors on where it matters whether it runs 11K or 47K operations a second? And that's just the simple micro-benchmark (which again does not reflect real world usage). In your complex benchmark, Chalk is much closer.

I just really like micro-optimization. I believe that we are in “Slow Web” because we are systematically moved performance optimizations for later.

I prefer Telegram’s (chat) approach when performance is an initial feature of the project,

Even when performance optimization is not visible, I want [by my projects] to promote performance-first thinking. For the case of Nano ID or Nano Colors, I am thinking that it is some sort of art: If we are thinking about performance of this smallest part, we will think about performance of thw whole system.

**Comment by nin-jin** [comment]
Hey, guys, stop war, make peace. Let's use [`felt-pen`](https://github.com/nin-jin/felt-pen) instead! It's:

- Completely original implementation instead of disputes about authorship of Nano Colors.
- Also haven't dependencies.
- Script size only 1kb (50sloc) instead of 3kb (140sloc) of Nano Colors.
- Package size is only 9kb instead of 16kb of Nano Colors.
- Also supports ESM, CommonJS, TypeScript.
- Also auto detects TTY.
- Also fast as Nano Colors.

Actual benchmarks:

### Calling

```
chalk          1,192,251 ops/sec
cli-color        130,947 ops/sec
ansi-colors      683,415 ops/sec
kleur          1,352,980 ops/sec
kleur/colors   1,383,478 ops/sec
colorette      1,503,513 ops/sec
felt-pen       1,414,047 ops/sec
nanocolors     1,530,478 ops/sec
```

### Loading

```
chalk         13.964 ms
cli-color     98.851 ms
ansi-colors    3.832 ms
kleur          7.182 ms
kleur/colors   1.469 ms
colorette      2.221 ms
felt-pen       1.424 ms
nanocolors     2.494 ms
```

**Comment by HyperHCl** [comment]
@nin-jin It's good fun y'all are having here! But...
 
* Env var logic for TTY not available.
* Uppercase for background name arguably confusing. I respect the choice though.
* Cannot set both fg and bg due to use of inverse for bg. It is an otherwise very cool idea.
* What's a `prefix`, `suffix`, and `postfix`?
* Very cool use of `if(`.

**Comment by nin-jin** [comment]
> Env var logic for TTY not available.

There is no need for it.

> Uppercase for background name arguably confusing. I respect the choice though.

Uppercase for more coloring, and lowercase for less coloring. It's very useful mental model. Actually it's not about background and foreground. Because my lib doesn't allow change foreground color on the colored background for readability purposes.

**Comment by mindplay-dk** [comment]
If you were to benchmark a babel/webpack/rollup run for any difference contributed by swapping color-coding libraries, I'm willing to bet, it's not even going to be measurable - if switching colors on the console contributed in any substantial way to the overall performance of these extremely large and complex tools, with all their parsers, compilers, source-maps, and what-have-you, there would have to be something *terribly* wrong with the performance of the color-coding library.

There is nothing that suggests that's the case.

To reiterate: this doesn't solve any problem.

**Comment by bakkot** [comment]
> No one has complained to us about Chalk's performance since we fixed it in Chalk v3 (2 years ago), but we are always open to working on improving performance.

Babel is still [mostly on Chalk v2](https://github.com/babel/babel/blob/767b72f9c2f49f82049d15ef8eb10094b0164f9a/package.json#L46), so perhaps it would be more straightforward just to upgrade it.

**Comment by Qix-** [comment]
> so perhaps it would be more straightforward just to upgrade it.

A cursory look through a Github search for `chalk` in this repository reveals that there shouldn't be any breakage by simply bumping the version. If someone would like to PR, feel free to ping me directly if any issues arise. I'll respond ASAP.

**Comment by JLHwung** [comment]
The blocker of chalk upgrade is that Babel 7 still supports Node 6. We can upgrade to Chalk 4 for Babel 8.

**Comment by ludofischer** [comment]
> If someone would like to PR, feel free to ping me directly if any issues arise. I'll respond ASAP.

I'm afraid you could at most bump Chalk to version 3 since I believe for now since Babel 7 still supports Node.js 8  (at least requiring Node.js 10 has only been merged for Babel 8 (see the[Babel release plan](https://github.com/babel/babel/issues/10746))

**Comment by dcastil** [comment]
Now that @sindresorhus released [yoctocolors](https://github.com/sindresorhus/yoctocolors) which is

> 4x faster and 5x smaller than nanocolors ([Source](https://github.com/sindresorhus/yoctocolors#features))

I think there‘s no argument left anymore to merge this. 😄

**Comment by shellscape** [comment]
> If you were to benchmark a babel/webpack/rollup run for any difference contributed by swapping color-coding libraries, I'm willing to bet, it's not even going to be measurable - if switching colors on the console contributed in any substantial way to the overall performance of these extremely large and complex tools, with all their parsers, compilers, source-maps, and what-have-you, there would have to be something _terribly_ wrong with the performance of the color-coding library.

Rollup team member and kinda-rollup-plugins-repo-lead here. This debate is conversational yak shaving. Rollup core wouldn't measurably be effected by ANSI color rendering on the CLI. But yeah we get these kinds of PRs as well. First it was `chalk` → `turbocolor` (https://github.com/rollup/rollup/pull/2339), then it was `turbocolors` → `ansi-colors` (https://github.com/rollup/rollup/pull/2412), then it was `turbocolor` → `colorette` (https://github.com/rollup/rollup/pull/4114). _This shit never ends on big projects._ Note the reasons in the first PR from @jorgebucaran for why Rollup should adopt `turbocolor` are the same arguments that @ai is making now. I don't know if that matters, but it's there. The primary reason that Rollup accepted the PR from @jorgebucaran for `turbocolor` was it reduced the shipped size of Rollup (as Rollup uses Rollup on itself before publishing to NPM). 

_(As a non-biased disclaimer I personally use `chalk` for all-the-things, I use `nanoid` like it's going out of style, and have had wonderful open source interactions with both @sindresorhus and @ai - they're both wonderful people with nothing but the best intentions and imho both have heaps of humility and very little ego.)_

**Comment by TheMarkie** [comment]
@ai @sindresorhus @nin-jin hey i just created a new text file and copypasted your code over into my project, tks 4 da open source, no credit given none due y'all retarded lmaoooooooooooooooooooooo yo check out my new npm package 8===========D~~~~~~

**Comment by SimenB** [comment]
This thread is just getting silly - I'll unsubscribe so I avoid getting it anywhere in my feed.

@nicolo-ribaudo I'd love to figure out the Jest bug - feel free to ping me from some other issue (or create and issue in Jest's tracker and tag me)

**Comment by benjamingr** [comment]
@SimenB - I might have ended up here after everyone but I only see leftover discussion and people making compelling arguments and working out conflict constructively. The communication from the project and from chalk/ colorette / nanocolors maintainers (at least not deleted) seems very reasonable and even though people are disagreement they are having a conversation.

I don't have a horse in the "what library to use here?" race - I just wanted to provide positive feedback about how the discussion has been going with everyone respecting each other and the project and everyone trying to communicate constructively even though things got heated.

Honestly I am even considering bringing this up to the Node.js moderation team so we can study how the conflict was successfully de-escalated...

**Comment by Qix-** [comment]
> seems very reasonable and even though people are disagreement they are having a conversation.

Just some context for anyone coming in late:

There were a few comments that were objectively abusive. The author has since been banned from Github and thus the comments have been deleted. I believe SimenB was referring to those, not the good-faith discussion prior. Try not to mis-interpret his comment as such 🙃

**Comment by fedeci** [comment]
I ask again to everyone that will come in the future to keep the thread clean for PR-related comments, please.

**Comment by SimenB** [comment]
> @SimenB - I might have ended up here after everyone but I only see leftover discussion and people making compelling arguments and working out conflict constructively.

The comment I was referring to has been deleted, so mine looks a bit out of context. That said, I'll once again unsubscribe from this PR - please don't tag me again here (GH seemingly has no option for muting a thread) ❤️

(The entire PR can probably be closed as the module has been deprecated: https://github.com/ai/nanocolors/commit/bb8e666e37ea5f452144aadf49eb4ebe3c0ce885)

**Comment by ai** [comment]
@SimenB I deprecated `nanocolors` in favour of `picocolors`, which is smaller, faster, and out of conflict.

Give me a day, I will choose this PR and create another one.

**Comment by SimenB** [comment]
I don't want to be subscribed to this issue, _please_ don't tag me

**Comment by mindplay-dk** [comment]
Another 0.x package released a week ago?

Please copy and paste all my previous comments.

(and if you hadn't seen it, the maintainer of that package makes fun of you @ai personally, in the README.)

This whole story now a farce - over a change that **nobody needs or asked for**.

**Comment by ehoogeveen-medweb** [comment]
> (and if you hadn't seen it, the maintainer of that package makes fun of you @ai personally, in the README.)

Perhaps this clarification is unnecessary, but given all the earlier drama: the comment in that package is just poking fun - the interactions between the authors on Twitter and in pull requests have been nothing but constructive.

**Comment by alonronin** [comment]
In Nano Colors repo they stated that the lib is deprecated and suggest using [Pico Colors]([https://github.com/alexeyraspopov/picocolors) instead:

> DEPRECATED: Use picocolors instead. It is 3 times smaller and 50% faster.

**Comment by ai** [comment]
All my projects was moved to `picocolors`. Let’s wait for a month or so to be sure that there is not known issues and project can be used safely.

I am closing the PR for now.
